### PR TITLE
Coverage

### PR DESCRIPTION
**Pull request by Lawrence37**
@aalmanza1998 Please check my changes to course_management.py.

### CODE REVIEW

File: ngshare_exchange/fetch_feedback.py [hunk 1]
```diff
@@ -1,42 +1,19 @@
...
-)
-from nbgrader.utils import parse_utc
-
 
 class ExchangeFetchFeedback(Exchange, ABCExchangeFetchFeedback):
     def init_src(self):
         if self.coursedir.course_id == '':
             self.fail('No course id specified. Re-run with --course flag.')
 
-        if self.coursedir.student_id != '*':
```
Comment (aalmanza1998) [inline ngshare_exchange/fetch_feedback.py:25]: Why is this removed?

File: ngshare_exchange/release_assignment.py [hunk 2]
```diff
@@ -112,14 +95,12 @@ def assignment_exists(self):
                             self.coursedir.assignment_id
                         )
                     )
-                    return True
```
Comment (aalmanza1998) [inline ngshare_exchange/release_assignment.py:115]: will self.fail ensure we return?

File: ngshare_exchange/tests/test_course_management.py [hunk 3]
```diff
@@ -344,14 +523,139 @@ def test_add_students_parsing(self, capsys):
                 'Missing column {} in {}.'.format('student_id', f.name) in out
             )
 
+        self._mock_add_students()
         with tempfile.NamedTemporaryFile() as f:
             f.write(b'student_id,first_name,last_name,email\n')
-            f.write(b',jane,doe,jd@mail.com')
+            f.write(b"sid1,jane,doe,jd@mail.com\n")
```
Comment (aalmanza1998) [inline ngshare_exchange/tests/test_course_management.py:529]:
```suggestion
            f.write(b'sid1,jane,doe,jd@mail.com\n')
```

File: ngshare_exchange/tests/test_course_management.py [hunk 4]
```diff
@@ -344,14 +523,139 @@ def test_add_students_parsing(self, capsys):
...
+            f.write(b',jane,doe,jd@mail.com\n')
+            f.write(b"sid2,john,perez,jp@mail.com\n")
```
Comment (aalmanza1998) [inline ngshare_exchange/tests/test_course_management.py:531]:
```suggestion
            f.write(b'sid2,john,perez,jp@mail.com\n')
```

File: ngshare_exchange/tests/test_course_management.py [hunk 5]
```diff
@@ -196,6 +255,85 @@ def test_add_student(self, capsys):
...
+
+    def test_add_students_db(self, capsys, tmpdir_factory):
+        tmp_dir = tmpdir_factory.mktemp(self.course_id)
+        os.chdir(tmp_dir)
+        self._add_empty_gradebook(tmp_dir)
+        self._mock_add_students()
+        with tempfile.NamedTemporaryFile() as f:
+            f.writelines(
+                [
+                    b"student_id,first_name,last_name,email\n",
```
Comment (aalmanza1998) [inline ngshare_exchange/tests/test_course_management.py:310]:
```suggestion
                    b'student_id,first_name,last_name,email\n',
```

File: ngshare_exchange/tests/test_course_management.py [hunk 6]
```diff
@@ -196,6 +255,85 @@ def test_add_student(self, capsys):
...
+                    b"sid1,jane,doe,jd@mail.com\n",
```
Comment (aalmanza1998) [inline ngshare_exchange/tests/test_course_management.py:311]:
```suggestion
                    b'sid1,jane,doe,jd@mail.com\n',
```

File: ngshare_exchange/tests/test_course_management.py [hunk 7]
```diff
@@ -196,6 +255,85 @@ def test_add_student(self, capsys):
...
+                    b"sid2,john,perez,jp@mail.com\n",
```
Comment (aalmanza1998) [inline ngshare_exchange/tests/test_course_management.py:312]:
```suggestion
                    b'sid2,john,perez,jp@mail.com\n',
```

File: ngshare_exchange/tests/test_course_management.py [hunk 8]
```diff
@@ -233,6 +371,22 @@ def test_add_students(self, capsys, tmp_path):
...
 
+    def test_add_students_unsuccessful(self, capsys, tmp_path):
+        self._mock_add_students_unsuccessful()
+        with tempfile.NamedTemporaryFile() as f:
+            f.writelines(
+                [
+                    b"student_id,first_name,last_name,email\n",
+                    b"sid1,jane,doe,jd@mail.com\n",
+                    b"sid2,john,perez,jp@mail.com\n",
+                ]
```
Comment (aalmanza1998) [inline ngshare_exchange/tests/test_course_management.py:382]:
```suggestion
                    b'student_id,first_name,last_name,email\n',
                    b'sid1,jane,doe,jd@mail.com\n',
                    b'sid2,john,perez,jp@mail.com\n',
                ]
```

File: ngshare_exchange/fetch_feedback.py (hunk 1 above)
Comment (Lawrence37) [reply ngshare_exchange/fetch_feedback.py:25]: The purpose of this block is to set `student_id`, but the variable is never used.

File: ngshare_exchange/release_assignment.py (hunk 2 above)
Comment (Lawrence37) [reply ngshare_exchange/release_assignment.py:115]: `self.fail` throws `ExchangeError`.

File: ngshare_exchange/list.py [hunk 9]
```diff
@@ -456,6 +455,7 @@ def nb_key(nb):
                 info['notebooks'].append(nb_info)
 
             if info['status'] == 'submitted':
+                assert not self.remove
```
Comment (Lawrence37) [inline ngshare_exchange/list.py:458]:
```suggestion
```
